### ISSUE TITLE
Uneven word alignment

### ISSUE THREAD

**Issue by zephster** [chrissimpkins/Hack#183]
Hi! Love the font, but this issue has been bugging me for a long time.

The alignment of bolded words seems to differ than non-bold words, and this only seems to happen on font sizes <= 10. Here's an example, in SublimeText 3:

Font size <=10: ![fontsize10](http://i.imgur.com/2LxVqY4.png)
Font size >=11: ![fontsize11](http://i.imgur.com/wE1PWLa.png)
(thead vs theads, just as an example)

Could this be caused by something in SublimeText, or is it an issue with the Hack font itself?
Thanks!

**Comment by chrissimpkins** [comment]
Thanks for the report Brandon.  This is not something that I’ve come across before.  What platform are you on and are you using the current release of the Hack fonts?

**Comment by zephster** [comment]
Windows 7 (Win10 at home, issue exists there as well), and I'm using the current release.

**Comment by chrissimpkins** [comment]
Thank you.  What sizes below 10 have you tried?

**Comment by zephster** [comment]
<=4: ![4](http://i.imgur.com/20FtdP0.png) - appears even, but too small to use
5-10: ![5](http://i.imgur.com/PBEXhnO.png) - appears uneven
\>=11: ![11](http://i.imgur.com/nTD1jtG.png) - appears even, but too large to use (in my opinion)

**Comment by chrissimpkins** [comment]
Thank you very much.  I am going to be doing some work with the hinting this weekend and will have a look.  We shouldn't be causing a shift off of the baseline with the hinting but let me check it to confirm.  What version of Sublime Text 3 are you using?  I believe that there was a new release in the last month or two.   I will check their release notes as well.  I don't think the problem lies with ST (and we can confirm if you would be willing to try it in another editor).

**Comment by zephster** [comment]
I'm using version 3107, but the issue's been present in every previous version I've used.

I installed Atom and tried it there, and the text does not appear to ever be un-aligned, at any font size.

However, I think that's just because Atom is not bolding the tag like ST3 does.
![](http://i.imgur.com/xGv3Qtv.png)

**Comment by chrissimpkins** [comment]
That is helpful.  Thank you!

**Comment by chrissimpkins** [comment]
I can verify that the hinted bold glyphs are properly aligned on the baseline (the bottom horizontal black line) across 8 - 14 ppem (and tested below that to 6 ppem however not shown in this animated gif):

![baseline-bold](https://cloud.githubusercontent.com/assets/4249591/13558470/4ad358ea-e3d3-11e5-8f06-c1b387556cce.gif)

and this animation shows alternate bold glyph displays with and without hints:

![hint-on-off](https://cloud.githubusercontent.com/assets/4249591/13558489/af089816-e3d3-11e5-8286-bf56c0fd2624.gif)

This is the case for the regular set too.  From what I can tell, there should not be a pixel difference in height between the glyphs in these two sets.  I'm not certain what is causing this issue.  Possible to try this in another editor with a syntax highlighter that includes bold elements to see if this ?

**Comment by zephster** [comment]
In Atom, I tried out a few different color schemes and manually edited them to have some bold tags. The issue didn't seem to happen, at any font size, for any color scheme.

In Sublime, the issue's present for every color scheme, at the previously discussed font sizes.

I'm going to install a few more editors/IDEs and see what happens.

Is there any other info I can provide for you?
Thanks for the help!

**Comment by chrissimpkins** [comment]
Interesting.  I don’t know why it is happening in ST and not in Atom.  I reviewed the changelog for the most recent ST 3 update and it appears that they made changes to their syntax highlighter approach; however, you say that this occurred in previous versions.  Might be worth a note on the ST forums to see if anyone on that side of the issue has any thoughts.  Post in here when you have a chance to try the other editors so that we know if this seems to be a problem elsewhere.    Thanks much!

**Comment by zephster** [comment]
I made a post on the ST forums [here](https://forum.sublimetext.com/t/uneven-font-alignment/18415).

I did some more testing:
Win10 VM, Eclipse, seemingly no alignment issues.
Win10 VM, standalone ST 2.0.2, no packages installed: no alignment issues.
Win10 VM, standalone ST3 build 3107 (the latest), no packages installed: no alignment issues.
Win7, standalone ST3 build 3107, no packages installed: alignment issues.

I also took the entire standalone 3107 directory from Win10 VM, which had no alignment issues, transferred everything to the host (Win7), and had alignment issues.

I'll try a fresh (standalone, no packages) setup when I get home, where I have Win10 as the native OS. The installed version of ST3 on my home computer also has the alignment issues, so I am assuming a fresh download will also suffer from this problem. I'll comment later when I find out for sure.

**Comment by chrissimpkins** [comment]
So it sounds like we can currently determine that this is an editor dependent issue (in Sublime Text) and that it occurs in native Win 7 and Win 10 OS's with the Sublime Text 3 editor.  Interestingly, it does not occur in VM's with either ST 2 or 3.  

One of the responders in your ST forum thread suggested that there have been other reports of a similar problem with other fonts.  Perhaps explore this in a bit more detail?  If you come up with more information we would be very interested to hear about it.  I don't know that there is anything at the moment that I can offer to solve it but we should leave this issue report open for others who come across a similar problem.

**Comment by chrissimpkins** [comment]
@zephster Any updates on this issue Brandon?

**Comment by zephster** [comment]
Unfortunately not. Activity in the ST3 forum thread died out, and I couldn't figure out any new meaningful information. It's something I'll just have to live with haha. I do appreciate the help, though!

**Comment by chrissimpkins** [comment]
If you have any interest we just released a new line spacing tool that adjusts the default line spacing in the fonts.  May try to nudge it a bit across all of the fonts.  Only other thing that I can think of.  If you want to give it a try the link is on our readme.

**Comment by chrissimpkins** [comment]
@zephster Brandon, do you mind trying an install of our new fonts with the new Windows installer to see if it addresses this issue?  There is no change in the fonts that should address this, but the installer itself might. We've seen correction of some unusual problems on Win systems when users upgrade with the tool rather than use the common install approach on the Windows platform.  Details here:

https://github.com/chrissimpkins/Hack/blob/master/README.md#windows

cc:@texhex

**Comment by zephster** [comment]
Oh snap. That seemed to do the trick! I'll post later to confirm if it
works on my home computer too, but it definitely worked for me here at the
office.

If you're ever in south Florida, I owe you a drink or three. Seriously.

**Comment by chrissimpkins** [comment]
Fantastic!  Glad to hear that it worked.  This was all @texhex 's work.  Tex, you traveling to S Florida anytime soon.  Free beer.  :)

Let us know if you still see these issues on your other system.

**Comment by texhex** [comment]
@chrissimpkins Well, that will be most expensive beer I ever had and the most confusing discussion with customs at the airport ever. 
C: Your reason for visiting the United States?
Me: To get my promised free drink.
C: ...?

**Comment by zephster** [comment]
you: "Sir, you don't understand. I helped restore this internet strangers sanity, and he promised me _good_ beer, not just some Bud Light garbage!"


Seriously though, thanks again for your help @chrissimpkins, and for the installer @texhex. I know this seems like a minor issue - and it is - but when you work in a text editor and stare at code all day something like this really stands out, and to me it was infuriating!

**Comment by texhex** [comment]
@zephster 
C: "Internet stranger? Not Bud Light?!? Sir, would you please follow me to the back! (Grabs disposable gloves)".

You're welcome Brandon, good to hear it works and we fully understand that once you noticed something like this, it can drive you insane. I'm crossing my fingers it will also help for your computer at home.

**Comment by chrissimpkins** [comment]
I actually think that it is the lederhosen that raises red flags through customs :)

![340px-lederhose](https://cloud.githubusercontent.com/assets/4249591/14866571/56a18cb8-0c92-11e6-87c0-e5095b855abb.jpg)


> Seriously though, thanks again for your help @chrissimpkins

My pleasure Brandon, always happy to help.

**Comment by texhex** [comment]
@chrissimpkins Or my T-Shirt with the "Clean Diesel by VW" slogan and the three bags of sausages I try to smuggle... :)

**Comment by texhex** [comment]
@zephster If you time permits and using the installer has worked on both computers, it would be great if you could paste the contents of `C:\Program Files\Hack Windows Installer\Log-FontData.txt` (from both computers) here. This could help us to identify the root cause of this issue.

**Comment by chrissimpkins** [comment]
> @chrissimpkins Or my T-Shirt with the "Clean Diesel by VW" slogan and the three bags of sausages I try to smuggle... :)

That'll do it

**Comment by zephster** [comment]
@texhex Here's the log file from my work computer; it did also work on my home computer I just forgot to get the log from it before I left, sorry.

---BeforeInstallAction START---
Setup version: 1.1.2
Font version.: 2.020
Local time...: 2016-27-04 14:48
Fonts folder.: C:\Windows\Fonts
Dest folder..: C:\Program Files\Hack Windows Installer
---HASH CALCULATION---
Calculating hash for Hack-Bold.ttf
   File from setup: a7bb6faacd609145b55ed15ca238755544c03af5
   File in \fonts : -NOT FOUND-
Calculating hash for Hack-BoldItalic.ttf
   File from setup: c428004a2fe3570450c6d03442052b1a9989c58b
   File in \fonts : 358e098533b14d6aa783c99c91bea270d536f69c
Calculating hash for Hack-Regular.ttf
   File from setup: 664cfe2a64de1486c0ace8073ceeb6d9281e8b78
   File in \fonts : -NOT FOUND-
Calculating hash for Hack-Italic.ttf
   File from setup: efdae4b94858b98eab6dcf2cb8e3cc3d28263cc2
   File in \fonts : 83b073724a04b883e948e8a52f52386340a023fe
----------------------
Checking for differences between fonts in setup and this system
IsSetupFontSameAsInstalledFont(): Hack-Bold.ttf
   Hash values (Setup/Windows): a7bb6faacd609145b55ed15ca238755544c03af5 / -NOT FOUND-
   File is different, installation required
   Found difference for file Hack-Bold.ttf
   Installation required.
---BeforeInstallAction END---
IsSetupFontSameAsInstalledFont(): Hack-Bold.ttf
   Hash values (Setup/Windows): a7bb6faacd609145b55ed15ca238755544c03af5 / -NOT FOUND-
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-BoldItalic.ttf
   Hash values (Setup/Windows): c428004a2fe3570450c6d03442052b1a9989c58b / 358e098533b14d6aa783c99c91bea270d536f69c
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-Regular.ttf
   Hash values (Setup/Windows): 664cfe2a64de1486c0ace8073ceeb6d9281e8b78 / -NOT FOUND-
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-Italic.ttf
   Hash values (Setup/Windows): efdae4b94858b98eab6dcf2cb8e3cc3d28263cc2 / 83b073724a04b883e948e8a52f52386340a023fe
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-Bold.ttf
   Hash values (Setup/Windows): a7bb6faacd609145b55ed15ca238755544c03af5 / -NOT FOUND-
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-BoldItalic.ttf
   Hash values (Setup/Windows): c428004a2fe3570450c6d03442052b1a9989c58b / 358e098533b14d6aa783c99c91bea270d536f69c
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-Regular.ttf
   Hash values (Setup/Windows): 664cfe2a64de1486c0ace8073ceeb6d9281e8b78 / -NOT FOUND-
   File is different, installation required
IsSetupFontSameAsInstalledFont(): Hack-Italic.ttf
   Hash values (Setup/Windows): efdae4b94858b98eab6dcf2cb8e3cc3d28263cc2 / 83b073724a04b883e948e8a52f52386340a023fe
   File is different, installation required
  Changes detected, require reboot

**Comment by chrissimpkins** [comment]
@zephster thank you!

**Comment by texhex** [comment]
@zephster Thanks a lot and no issue. Good to hear it works now. The log reveals one interesting part:

The Hack font was partially installed to c:\Windows\Fonts (e.g. Hack Italic) but other types (especially Hack Regular) was *NOT*.

Calculating hash for Hack-Regular.ttf
File from setup: 664cfe2a64de1486c0ace8073ceeb6d9281e8b78
File in \fonts : -NOT FOUND-

Because the font registration does not require that a font exists in C:\Windows\Fonts, this might have been the root cause of all of this. 

However, as the log does not indicate to which files the registration pointed to, I'm just guessing here. I have created an [issue](https://github.com/source-foundry/Hack-windows-installer/issues/8) over at the Hack-Windows-Installer repository to include this information in the log.

Brandon, would it be OK for you to close this issue? Given that the issues where no font issue but installation issue and it was resolved using the new installer?

@chrissimpkins: Also OK for you?
